File: lab2/Dfa.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LFPC_lab2_final
{
    public class Dfa
    {
        private readonly int startState;
        private readonly Set<int> acceptStates;
        private readonly IDictionary<int, IDictionary<String, int>> trans;

        public Dfa(int startState, Set<int> acceptStates, IDictionary<int, IDictionary<String, int>> trans)
        {
            this.startState = startState;
            this.acceptStates = acceptStates;
            this.trans = trans;
        }

        public int Start { get { return startState; } }

        public Set<int> Accept { get { return acceptStates; } }

        public IDictionary<int, IDictionary<String, int>> Trans
        {
            get { return trans; }
        }

        public override String ToString()
        {
            return "DFA start=" + startState + "\naccept=" + acceptStates;
        }

        public void WriteDot(String filename)
        {
            TextWriter wr =
              new StreamWriter(new FileStream(filename, FileMode.Create,
                                              FileAccess.Write));
            wr.WriteLine("// Format this file as a Postscript file with ");
            wr.WriteLine("//    dot " + filename + " -Tps -o out.ps\n");
            wr.WriteLine("digraph dfa {");
            wr.WriteLine("size=\"11,8.25\";");
            wr.WriteLine("rotate=90;");
            wr.WriteLine("rankdir=LR;");
            wr.WriteLine("n999999 [style=invis];");    // Invisible start node
            wr.WriteLine("n999999 -> n" + startState); // Edge into start state

            // Accept states are double circles
            foreach (int state in trans.Keys)
                if (acceptStates.Contains(state))
                    wr.WriteLine("n" + state + " [peripheries=2];");

            // The transitions 
            foreach (KeyValuePair<int, IDictionary<String, int>> entry in trans)
            {
                int s1 = entry.Key;
                foreach (KeyValuePair<String, int> s1Trans in entry.Value)
                {
                    String lab = s1Trans.Key;
                    int s2 = s1Trans.Value;
                    wr.WriteLine("q" + s1 + " -> q" + s2 + " [symbol=\"" + lab + "\"];");
                }
            }
            wr.WriteLine("}");
            wr.Close();

        }
    }
}

File: lab2/Dfa.java
package annachiriciuc;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Map;
import java.util.Set;

public class Dfa {
    private Integer startState;
    private Set<Integer> acceptStates;
    private Map<Integer, Map<String, Integer>> trans;

    public Dfa(Integer startState,
               Set<Integer> acceptStates,
               Map<Integer, Map<String, Integer>> trans) {
        this.startState = startState;
        this.acceptStates = acceptStates;
        this.trans = trans;
    }

    public Integer getStart() {
        return startState;
    }

    public Set<Integer> getAccept() {
        return acceptStates;
    }

    public Map<Integer, Map<String, Integer>> getTrans() {
        return trans;
    }

    public String toString() {
        return "DFA start=" + startState + "\naccept=" + acceptStates
                + "\n" + trans;
    }

    public void writeDot(String filename) throws IOException {
        PrintWriter out = new PrintWriter(new FileWriter(filename));
        // Accept states are double circles
        for (Integer state : trans.keySet())
            if (acceptStates.contains(state))
                out.println("q" + state);

        // The transitions
        for (Map.Entry<Integer, Map<String, Integer>> entry : trans.entrySet()) {
            Integer s1 = entry.getKey();
            for (Map.Entry<String, Integer> s1Trans : entry.getValue().entrySet()) {
                String lab = s1Trans.getKey();
                Integer s2 = s1Trans.getValue();
                out.println("q" + s1 + " -> q" + s2 + " [weight=" + lab + "];");
            }
        }
        out.println("}");
        out.close();
    }
}

